tree/tree/test/TBasket.cxx

```diff
@@ -0,0 +1,146 @@
+
+#include "gtest/gtest.h"
+
+#include "TMemFile.h"
+#include "TTree.h"
+#include "TBranch.h"
+#include "TBasket.h"
+
+#include <vector>
+
+static const Int_t gSampleEvents = 100;
+
+void createSampleFile(TMemFile *&f)
+{
+   f = new TMemFile("tbasket_test.root", "CREATE");
+   ASSERT_TRUE(f != nullptr);
+   ASSERT_FALSE(f->IsZombie());
+
+   TTree t1("t1", "Simple tree for testing.");
+   ASSERT_FALSE(t1.IsZombie());
+   Int_t idx;
+   t1.Branch("idx", &idx, "idx/I");
+
+   for (idx=0; idx < gSampleEvents; idx++) {
+      t1.Fill();
+   }
+   t1.Write();
+}
+
+void verifySampleFile(TFile *f)
+{
+   TTree *tree = nullptr;
+   f->GetObject("t1", tree);
+   ASSERT_TRUE(tree != nullptr);
+
+   Int_t saved_idx;
+   tree->SetBranchAddress("idx", &saved_idx);
+
+   EXPECT_EQ(tree->GetEntries(), gSampleEvents);
+   for (Int_t idx = 0; idx < tree->GetEntries(); idx++) {
+      tree->GetEntry(idx);
+      EXPECT_EQ(idx, saved_idx);
+   }
+}
+
+// Basic "sanity check" test -- can we create and delete trees?
+TEST(TBasket, CreateAndDestroy)
+{
+   std::vector<char> memBuffer;
+
+   TMemFile *f;
+   createSampleFile(f);
+   f->Close();
+
+   Long64_t maxsize = f->GetSize();
+   memBuffer.reserve(maxsize);
+   f->CopyTo(&memBuffer[0], maxsize);
+
+   delete f;
+
+   TMemFile f2("tbasket_test.root", &memBuffer[0], maxsize, "READ");
+   ASSERT_FALSE(f2.IsZombie());
+   verifySampleFile(&f2);
+}
+
+// Create a TTree, pull out a TBasket.
+TEST(TBasket, CreateAndGetBasket)
+{
+   TMemFile *f;
+   createSampleFile(f);
+   ASSERT_FALSE(f->IsZombie());
+
+   TTree *tree = nullptr;
+   f->GetObject("t1", tree);
+   ASSERT_TRUE(tree != nullptr);
+   ASSERT_FALSE(tree->IsZombie());
+
+   TBranch *br = tree->GetBranch("idx");
+   ASSERT_TRUE(br != nullptr);
+   ASSERT_FALSE(br->IsZombie());
+
+   TBasket *basket = br->GetBasket(0);
+   ASSERT_TRUE(basket != nullptr);
+   ASSERT_FALSE(basket->IsZombie());
+
+   EXPECT_EQ(basket->GetNevBuf(), gSampleEvents);
+   verifySampleFile(f);
+
+   f->Close();
+   delete f;
+}
+
+TEST(TBasket, TestUnsupportedIO)
+{
+   TMemFile *f;
+   // Create a file; not using the createSampleFile helper as
+   // we must corrupt the basket here.
+   f = new TMemFile("tbasket_test.root", "CREATE");
+   ASSERT_TRUE(f != nullptr);
+   ASSERT_FALSE(f->IsZombie());
+
+   TTree t1("t1", "Simple tree for testing.");
+   ASSERT_FALSE(t1.IsZombie());
+   Int_t idx;
+   t1.Branch("idx", &idx, "idx/I");
+   for (idx=0; idx < gSampleEvents; idx++) {
+      t1.Fill();
+   }
+
+   TBranch *br = t1.GetBranch("idx");
+   ASSERT_TRUE(br != nullptr);
+
+   TBasket *basket = br->GetBasket(0);
+   ASSERT_TRUE(basket != nullptr);
+
+   EXPECT_EQ(basket->fIOBits, 0);
+
+   // This tests that at least one bit in the bitset is available.
+   // When we are down to one bitset, we'll have to expand the field.
+   EXPECT_TRUE(static_cast<UChar_t>(~TBasket::kSupported));
+
+   basket->fIOBits = ~TBasket::kSupported;
```

Comment:
Shouldn't we have TBasket::GetIOBits() instead? (and/or TestIOBits(TBasket::EIOBits))

Reply:
Humm ... we do ... but I see ... this is an assignment here ... You could also do it by grabbing the offset of fIOBits via TClass/TDataMember.

Reply:
I want to avoid providing interfaces to use the IOBits infrastructure in this PR (aim is to keep backports simple -- no new API is introduced into older versions of ROOT if we backport this).

While the `TClass`/`TDataMember` technically could be done -- is it really more straightforward than just allowing the unit test to be a friend class?

Reply:
It is relatively straightforward but indeed not as straightforward as the friend class.

However I dislike the spurrious friend declaration .. and 'more' importantly the name of the class is actually dependent on the gtest implementation.  They may or may not change their naming algorithm in a future version.